Add architectural support for multi-GPU.

AshkanAliabadi opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #44017 Add Vulkan op Add.
* #44016 Add Vulkan Tensor factory.
* #44015 Add Vulkan Tensor.
* #45148 Add fence.
* #45037 Add image sampler.
* #44496 Redo Vulkan command and descriptor pools.
* **#44059 Add architectural support for multi-GPU.**
* #44317 Minor touchups.

Differential Revision: [D23820825](https://our.internmc.facebook.com/intern/diff/D23820825)

Commit:
Add architectural support for multi-GPU.

[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

AshkanAliabadi commented on aten/src/ATen/native/vulkan/api/Context.cpp:
```diff
@@ -8,221 +8,31 @@ namespace vulkan {
+    try {
+      const Adapter adapter = runtime()->select([](const Adapter& adapter) {
+        // Select the first adapter.
+        return true;
```
This lambda can later be replaced with something smarter to select the integrated GPU or the most computationally capable, etc.

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Common.h:
```diff
@@ -48,6 +49,12 @@ struct Resource;
 struct Shader;
 
+struct GPU final {
+  const Adapter* adapter;
```
Is it ok to have here pointer and not reference, will it be valid to have nullptr adapter?

AshkanAliabadi replied:
Great point, thanks.  Having a reference will make this class non-copyable.  I think in this particular case it might be a problem considering my use of this class elsewhere, but I will double check to see if that is the case.  You have a solid point there though that references are superior to pointers.

IvanKobzarev commented on aten/src/ATen/native/vulkan/api/Adapter.h:
```diff
@@ -0,0 +1,36 @@
+//
+
+struct Adapter final {
+  Runtime* runtime;
```
Thinking about if reference can be used here.

Could it be nullptr here for some cases?

AshkanAliabadi replied:
Great point, thank you.  Reference members make classes uncopyable, which is the reason behind my use of pointers, but I actually never checked to see if it is OK for Adapter to be non-copyable.  I'll make this change and see if I can make it work.  Thanks.

AshkanAliabadi replied:
To report back on this - I tried converting the pointers to references both here and on `api::GPU` but that makes them makes them non-move-assignable (can move construct though) besides non-copyable which means objects of these types cannot be returned from functions without mandatory copy elision introduced in C++-17.  :/

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."




[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."


Differential Revision: [D23820825](https://our.internmc.facebook.com/intern/diff/D23820825)

[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."


Differential Revision: [D23820825](https://our.internmc.facebook.com/intern/diff/D23820825)

[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."


Differential Revision: [D23820825](https://our.internmc.facebook.com/intern/diff/D23820825)

[ghstack-poisoned]

Commit:
Update on "Add architectural support for multi-GPU."


Differential Revision: [D23820825](https://our.internmc.facebook.com/intern/diff/D23820825)

[ghstack-poisoned]
